### PR TITLE
Improve performance for undo and for gathering cell metas

### PR DESCRIPTION
**Pull request by budnix**
### Context

~The added argument determines if the table renders cycle should be performed after setting new source data. This fixes the performance issues that occur in multiple method calls.~

It turned out that the above-mentioned argument is not necessary. It is enough to collect all changes and call the `setSourceDataAtCell` only once.

While developing I spotted that context menu for a bigger dataset (10000x10) is about 5 times slower in opening than in the 7.4.2. It was caused by slow cell meta gathering. It turned out that extending class using ES6+ syntax (babel transformation) causes performance degradation (https://github.com/handsontable/handsontable/blob/feature/issue-6868/src/dataMap/metaManager/utils.js#L49). So, I changed the syntax to use old function style with changing its prototype.

After some local tests, I can confirm that the context menu shows 1.25 times faster than 7.4.2, and `undo` operation (after removing columns) is ~1.2 faster than 7.4.2. 

Additionally I've added a code that refreshes the editor value when the data source was changed. The bug was fixed for the `setDataAtCell` method so the same applied for the source data.

### How has this been tested?

Added tests for editor value refreshing and test for `render` call.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6868
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -2394,22 +2393,35 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
-        input[i][0],
-        input[i][1],
-        dataSource.getAtCell(input[i][0], input[i][1]),
-        input[i][2],
-      ]);
+    const isThereAnySetSourceListener = this.hasHook('afterSetSourceDataAtCell');
+    const changesForHook = [];
+
+    if (isThereAnySetSourceListener) {
+      arrayEach(input, ([changeRow, changeProp, changeValue]) => {
```
Comment (wszymanski) [inline src/core.js:2400]: Do you use helpers :)? Maybe native `Array.each` will be faster?

File: src/dataMap/metaManager/utils.js [hunk 2]
```diff
@@ -40,7 +40,15 @@ export function expandMetaType(type, metaObject) {
  * @returns {ColumnMeta} Returns constructor ready to initialize with `new` operator.
  */
 export function columnFactory(TableMeta, conflictList = []) {
-  class ColumnMeta extends TableMeta {}
+  // Do not use ES6 "class extends" syntax here. It seems that the babel produces code
+  // which drastically decreases the performance of the ColumnMeta class creation.
```
Comment (wszymanski) [inline src/dataMap/metaManager/utils.js:44]: Nice investigation!

File: src/plugins/undoRedo/undoRedo.js [hunk 3]
```diff
@@ -428,39 +427,48 @@ UndoRedo.RemoveColumnAction.prototype.undo = function(instance, undoneCallback)
...
-  });
 
-  let sortedHeaders = [];
-  sortedHeaders = arrayMap(this.headers, sortByIndexes);
+  for (let rowIndex = 0; rowIndex < removedDataLength; rowIndex++) {
+    sortedData.push(arrayMap(this.data[rowIndex], sortByIndexes));
+  }
 
+  const sortedHeaders = arrayMap(this.headers, sortByIndexes);
+  const isFormulaPluginEnabled = instance.getPlugin('formulas') ? instance.getPlugin('formulas').enabled : false;
```
Comment (wszymanski) [inline src/plugins/undoRedo/undoRedo.js:438]: I don't know whether it is cleaner 🤔.

```suggestion
  const isFormulaPluginEnabled = instance.getPlugin('formulas')?.enabled || false;
```

File: src/plugins/undoRedo/undoRedo.js [hunk 4]
```diff
@@ -428,39 +427,48 @@ UndoRedo.RemoveColumnAction.prototype.undo = function(instance, undoneCallback)
...
   const changes = [];
 
   instance.alter('insert_col', this.indexes[0], this.indexes.length, 'UndoRedo.undo');
 
-  rangeEach(this.data.length - 1, (i) => {
-    const row = instance.getSourceDataAtRow(i);
+  arrayEach(instance.getSourceDataArray(), (rowData, rowIndex) => {
```
Comment (wszymanski) [inline src/plugins/undoRedo/undoRedo.js:443]: What about native `Array.each`?

File: src/core.js (hunk 1 above)
Comment (budnix) [reply src/core.js:2400]: `arrayEach` is more stable :) There is no policy about the usage of the native vs helpers. I'll leave as it is.

File: src/plugins/undoRedo/undoRedo.js (hunk 4 above)
Comment (budnix) [reply src/plugins/undoRedo/undoRedo.js:443]: `arrayEach` is more stable :) There is no policy about the usage of the native vs helpers. I'll leave as it is.

File: src/plugins/undoRedo/undoRedo.js (hunk 3 above)
Comment (budnix) [reply src/plugins/undoRedo/undoRedo.js]: Fixed within https://github.com/handsontable/handsontable/pull/6930/commits/fea3b26e978e4beeb77ca87d411f21d3e5e3edf9 👍

**Pull request merged**